R/derive_interp_records.R

```diff
@@ -44,73 +45,102 @@
   }
 
-  arrange(dataset, !!!by_vars, AGE)
+  # Ensure to have AGE unit in "Days"
+  if (any(toupper(dataset$AGEU) != "DAYS")) {
+    cli_abort("The Age Unit (AGEU) from the input dataset must be in 'DAYS'")
+  }
+
+  # Sort the data for the interpolation
+  dataset <- arrange(dataset, !!!by_vars, AGE)
 
-  # Ensure to have unique combination when by_vars is not defined
-  metadata_vars <- c("AGE", "L", "M", "S")
+  # Define the metadata variables to be interpolated
+  interp_vars <- c("AGE", "L", "M", "S")
   if (parameter == "BMI") {
-    metadata_vars <- append(metadata_vars, c("P95", "Sigma"))
+    interp_vars <- append(interp_vars, c("P95", "Sigma"))
   }
 
-  if (is.null(by_vars)) {
-    other_data_vars <- names(dataset %>% select(-all_of(metadata_vars)))
+  # Ensure the uniqueness of records to interpolate
+  signal_duplicate_records(dataset, by_vars = exprs(!!!by_vars, AGE))
 
-    nb_occ <- nrow(dataset %>%
-      group_by(across(all_of(other_data_vars))) %>%
-      slice(1) %>%
-      ungroup())
-
-    if (nb_occ > 1) {
-      cli_abort(paste0(
-        "The combination of ", paste(other_data_vars, collapse = ", "),
-        " must be unique. Please define `by_vars` otherwise."
-      ))
-    }
+  # Define the non-interpolated variables and keep the corresponding unique records
+  non_interp_vars <- setdiff(names(dataset), c(interp_vars, by_vars))
+  if (length(non_interp_vars) > 0) {
+    non_interp_dataset <- dataset %>%
+      select(
+        map_chr(replace_values_by_names(by_vars), as_label),
+        c(all_of(non_interp_vars), "AGE")
+      ) %>%
+      unique()
   }
 
   # Linear interpolation
   fapp <- function(v, age) {
     approx(age, v, xout = seq(min(age), max(age)))$y
   }
 
+  # Apply LOCF to non-interpolated variables
+  apply_locf <- function(x) {
+    na.locf(x, na.rm = FALSE)
+  }
+
   # Apply the function within each group and combine the results
   if (is.null(by_vars)) {
-    dataset <- dataset %>%
+    interp_dataset <- dataset %>%
       reframe({
         age <- AGE
-        x <- lapply(select(., all_of(metadata_vars)), fapp, age = age)
+        x <- lapply(select(., all_of(interp_vars)), fapp, age = age)
         as.data.frame(do.call(bind_cols, x))
       }) %>%
       filter(!is.na(AGE))
   } else {
-    dataset <- dataset %>%
-      group_by(across(!!!syms(map(replace_values_by_names(by_vars), as_label)))) %>%
+    interp_dataset <- dataset %>%
+      group_by(across(map_chr(replace_values_by_names(by_vars), as_label))) %>%
       reframe({
         age <- AGE
-        x <- lapply(across(all_of(metadata_vars)), fapp, age = age)
+        x <- lapply(across(all_of(interp_vars)), fapp, age = age)
         as.data.frame(do.call(bind_cols, x))
       }) %>%
       ungroup() %>%
       filter(!is.na(AGE))
   }
+
+  # Merge non-interpolated variables (if any) back into the interpolated dataset
+  if (length(non_interp_vars) > 0) {
```

Comment:
This condition and the `else` branch should be remove because `non_interp_vars` contains at least `AGEU`.